### PR TITLE
Shapes: a strong-call-by-need evaluator

### PR DESCRIPTION
**Pull request by gasche**
This PR sits on top of #10808; instead of a strong call-by-value evaluator, it implements a strong call-by-need evaluator following the approach of @Ekdohibs in https://github.com/ocaml/ocaml/pull/10808#issuecomment-987683465 .

The code, surprisingly, ends up somewhat simpler. (Well, maybe @Ekdohibs would know how to write a better strong-call-by-value evaluator than my own.)

But there is one potential issue: to perform memoization (using a hashtable) on environments that contain lazy thunks, I wrap the thunks in an object (`'a Lazy.t` becomes `< force : 'a >`) so that it becomes hashable. This works fine, but I don't remember if using (immediate) objects in the type-checker is strictly forbidden or not, for bootstrapping reasons. (We already use exceptions and polymorphic variants, but...) (Note that shape reduction runs unconditionally, even when `-bin-annot` is not passed. Otherwise I guess it would be fine, but it's unclear how to structure the code to only run conditionally.)

### CODE REVIEW

**Comment by garrigue** [comment]
About the question of using objects in the type checker: I think the main problem is that objects use a runtime (camlinternalOO.ml), so that changes in this runtime may have an impact on the bootstrap.
For safety, it would indeed be better if shape reduction could be disabled easily.

Actually, it is a bit surprising that using objects is the only solution to create values with unique identifiers recognized by the comparison functions. Shouldn't we have those, maybe by allowing the class pointer in object blocks to be `Val_unit` ?

**Comment by voodoos** [comment]
Thanks again for this PR @gasche.

I just ran the same benchmarks as I did for PR #10796 and here are the results:

### OCaml 4.14 without shapes (based on 0f152a868fca5f7801ffc3be9fb881f44f852a41)
| Command | Mean [s] | Min [s] | Max [s] | Relative |
|:---|---:|---:|---:|---:|
| `dune build @install -j 1` | 45.209 ± 0.278 | 44.787 | 45.590 | 1.00 |

### This PR
| Command | Mean [s] | Min [s] | Max [s] | Relative |
|:---|---:|---:|---:|---:|
| `dune build @install -j 1` | 46.167 ± 0.267 | 45.787 | 46.573 | 1.00 |

Difference between the size of the `cmt` files is negligible.

Here are the timings for the files which show the biggest differences:

cm_file | 414-no-shapes | 414-shape-strong-cbn
-- | -- | --
irmin_git.cmx | 1.228 (±0.001) | 1.315 (±0.056)
irmin_git.cmo | 1.345 (±0.022) | 1.427 (±0.004)
irmin_unix__Xgit.cmx | 0.973 (±0.092) | 1.029 (±0.011)
irmin_mirage_git.cmx | 1.023 (±0.016) | 1.067 (±0.020)
irmin.cmx | 0.423 (±0.013) | 0.458 (±0.009)
irmin.cmo | 0.402 (±0.003) | 0.432 (±0.009)
irmin_containers__Counter.cmx | 0.118 (±0.003) | 0.144 (±0.003)
irmin_mirage_git.cmo | 1.050 (±0.038) | 1.073 (±0.002)

As expected, the difference even in the worst case is much less than before.

Moreover I upgraded Merlin to 4.14 and ported our new locate prototype along with this PR: all tests pass succesfuly and a bunch of previously incorrect ones are now correct as expected.

Finally @trefis is going to run Merlin locate with these changes at random positions in Irmin to compare the results and speed of this version with regard to the last release.

**Comment by gasche** [comment]
Thanks! I wonder if you also measured the `.cmt` file size? (We *suppose* that the strong normal forms are going to stay reasonably-sized, but in theory some of them could be large in size even in presence of memoization/hashconsing/sharing.)

**Comment by voodoos** [comment]
> wonder if you also measured the `.cmt` file size?

In bytes:

cm_file | 414-no-shapes | 414-shape-strong-cbn
-- | -- | --
irmin.cmt | 2,537,692 | 2,552,510
irmin.cmti | 1,199,985 | 1,200,707

So the size difference is negligible in the case of Irmin.

**Comment by trefis** [comment]
> Finally @trefis is going to run Merlin locate with these changes at random positions in Irmin to compare the results and speed of this version with regard to the last release.

W.r.t. speed this version is a clear improvement over what merlin was doing before shapes (where it sometimes took several cases to answer), and also a bit better than the version with substitutions on those degenerate cases:
some graphs
(N.B. 414 is the version which doesn't use shapes, strong-cbn is the version of this PR, and subst-sharing the version of our PR)

On non degenerate cases this version appears slightly slower than the other one, but not in a way that's ever going to be noticeable by a user it seems.

So it seems we have our winner?

**Comment by trefis** [comment]
Apologies, I had *obviously* forgotten to build the `subst-sharing` cmts with the `subst-sharing` compiler.
With that mistake corrected, the results are even more in favor of this PR (vastly so):

some graphs

**Comment by gasche** [comment]
Thanks for the performance numbers!

Here is my current understanding:
- we know that strong reduction does result in size explosions on existing programs
- memoization is effective at making known explosions go away
- we don't know whether other existing programs still suffer from size-explosion in presence of memoization, but it's reasonable to wait until we meet one to worry about that
- the performance of this PR is adequate, it makes strong reduction in the compiler very reasonable in practice

I am curious about the performance different between (memoized) call-by-value and (memoized) call-by-need on this problem, that is between #10808 and #10825. If it's not too much work, could you rerun the last-percentile programs to see if they differ in performances? In theory strong call-by-need is faster in corner cases and constant-factor slower in general due to extra book-keeping, but my guess would be that memoization is the key aspect here and the two strategies are indistinguishable.

We still need to fix the objects-used-in-the-frontend issue. Possible solutions include:
1. writing our own hash function for shape terms and normal forms (but: writing a good and fast hash function is cumbersome)
2. offloading the lazy thunks to a dynamic array (or int-indexed hashtable), keeping only the indices in the structure
3. making sure that shape normalization does not run unless `-bin-annot` or `-dshape` are used

(1) is more boring than (2), which makes it a wiser choice. But I think that we should do (3) in any case; there is something fishy about the current state that deserves a discussion.

**Comment by xavierleroy** [comment]
> 1.  writing our own hash function for shape terms and normal forms (but: writing a good and fast hash function is cumbersome)
> 2.  offloading the lazy thunks to a dynamic array (or int-indexed hashtable), keeping only the indices in the structure
> 3.  making sure that shape normalization does not run unless -bin-annot or -dshape are used

I think 3 would suffice to avoid the risk of a bug in object compilation or in CamlinternalOO breaking the bootstrap, no need for more fancy approaches.

For 1, you could use a reference to a map instead of a hash table.  Then, you only need  to write a comparison function, which is easier than writing a hash function + an equality function.

**Comment by gasche** [comment]
> making sure that shape normalization does not run unless -bin-annot or -dshape are used

My understand is that, currently, shapes are constructed unconditionally (whether or not they will be stored on disk) and also normalized unconditionally.

I think It would be nice to construct and normalize them only conditionally. Three questions:
1. conditional construction
2. are shapes only stored in .cmt files?
3. conditional normalization

### Conditional construction?

The construction of shapes is currently interleaved with the type-checker logic, so we cannot make it conditional without an invasive change to compute them directly from the typedtree. I think the change would be desirable, but it's definitely not in scope for 4.14.

### Are shapes only used in .cmt files?

Currently, most typedtree structures and typing-environment entries contain shape information. This shape information might leak in `.cmo` and `.cmi` files, or in other `-dfoo` outputs; if this was to happen, then conditional normalization would be an issue (it would produce different outputs whether or not `-bin-annot` is passed).

My impression is that we may be safe (no leaking of shapes outside `.cmt`):
- it seems that environment summaries do not contain any shape information
- `-dtypedtree` ignores shape information
- the `shape` field of `Typedtree.implementation` is ignored unless `-bin-annot` or `-dshape`

I would welcome confirmation from @trefis on these points.

I must say that the non-obviousness of this all makes me a bit nervous, design-wise. Again, computing shapes outside the type-checker would make these questions go away.

### Conditional normalization

Currently we normalize shapes unconditionally. We could make normalization conditional, but there are two problems:
- with the current code design, things would be a bit weird: the typedtree would store shape information whose normalization status would depend on whether `-bin-annot` or `-dshapes` are used.
- if we were wrong, and some shapes are actually stored in `.cmi` or `.cmo` files, then conditional normalization can lead to painful bugs where different compiler flags result in different digests and broken build.

So we have to balance two risks against each other:
- unconditional normalization could lead to shape-size-explosion on some still-unknown programs
- conditional normalization could lead to painful digest bugs if it is enabled automatically

I think that unconditional normalization is the lesser evil here. (In any case, we assume that most developers use `-bin-annot` to get good Merlin feedback, so they will encounter blowups.) If we wanted to be *really* safe, we could provide an *opt-in* flag to disable shape normalization, but (1) this is work, (2) the user interface is ugly ("use this escape hatch if the compiler blows up!"), and (3) it may not make sense anymore in a couple compiler versions if we feel good about conditional normalization. I'm not planning to work on such a flag for now.

**Comment by gasche** [comment]
@xavierleroy that's a good point. I would expect performance overhead from using a map, but maybe not (in fact if shapes different quickly enough, comparison can traverse less of the structure than hashing), and in any case normalization is fast enough in practice. I'll try that.

**Comment by Ekdohibs** [comment]
What about using `type 'a thunk = int * 'a Lazy.t` ? The integer could be a fresh counter, and the hash or comparison function simply refer to that integer. I think this would be semantically equivalent to using the identity to compare objects, while not using objects.

**Comment by gasche** [comment]
Yes, this is what I am working on, but this is actually a lot of code to write (relatively to the whole evaluator): a comparison function for the `nf` and `local_env` datatypes, then two functor instantiations for the two memoization tables (one for `eval` and one for `read_back`), etc.

Note: the point is that `int * 'a Lazy.t` cannot be safely hashed by the runtime-provided hash function, so you have to write your own hash function for any datatype that contains values of this type, which is tedious. Wrapping the lazy value inside an an object avoided this problem by letting us reuse the standard hash, objects are hashed using their unique identifier only.

**Comment by alainfrisch** [comment]
FWIW, we (LexiFi) sometimes use the trick of defining a record type with an int (a fresh counter) as the second field, and forcing the tag values to be 252, thus allowing to use generic hashing/comparison on big data structures containing those records (with a physical equality semantics for them).  It could make sense to expose this hack as a proper module in the Stdlib, e.g. as reference cells compared with physical equality:

````ocaml
module PhysRef : sig
  type 'a t (* = private { mutable contents: 'a; uid: int } ?? *)
  val create: 'a -> 'a t
  val get: 'a t -> 'a
  val set: 'a t -> 'a -> unit
  (* val id: 'a t -> int ?? *)
end
````

**Comment by gasche** [comment]
I pushed two new commits to implement memoization using a Map and a user-written comparison functions. (This takes 103 lines of code, a 77% increase to the previous PR additions to typing/shape.ml.)

Note: I cannot observe any performance difference when building irmin.cmt.

**Comment by gasche** [comment]
Current status: I am satisfied by the implementation of both this PR #10825 (a memoized strong call-by-need (CBN) evaluator for shapes) and #10808 (a memoized strong call-by-value (CBV) evaluator for shapes). We probably need to merge either of them in 4.14 -- or go back on shapes for 4.14 altogether.

This requires a review from a maintainer (at least). Who is going to do the review work? I can be available for an online review meeting, and @voodoos might be interested in attending as well.

(The history of either PRs is almost-clean, and designed for commit-by-commit review. I want to remove the previous, uglier implementation of strong CBV that is present on both sides. I'm planning to do this very soon, so that the PR can comfortably be reviewed commit-by-commit.)

File: typing/shape.ml
Comment (Ekdohibs) [inline typing/shape.ml]: Maybe this function could associate `0` to `NLeaf` and numbers `1-7` in the right order for the constructors with arguments? That way, the compiler could optimize the function using simple arithmetic on the block's tag instead of a more complex computation.

Comment (gasche) [reply typing/shape.ml]: Correct (this was my intent, but I forgot that there was a constant constructor in the mix), but I don't think it matters performance-wise. I'll do the change anyway, as it makes the code more natural.

**Comment by gasche** [comment]
Over in https://github.com/ocaml/ocaml/pull/10808#issuecomment-1005255786, @Ekdohibs remarks:

> You could completely avoid the use of `Lazy.t` in the strong call-by-need evaluator as well using the type `(nf, t * env) result ref` where you would have a lazy value, since the code pointer used for the closure is always the same (reduce), and the important part is the value of the variables bound by this closure, i.e. a term and an environment. You could also add a `NLazy of t * env` constructor to the type `nf`, and simply use `nf ref` where a lazy value is needed, taking care to evaluate it.

I thought about this as well, but it does not actually work: replacing `'a lazy` by `'a option ref` prevents the hashing function from crashing, but memoization does not work as expected because the hash/comparison result will depend on the value of the reference: hashing is not stable with respect to completion of delayed computations. 

On the other hand, one thing that I think *would* work is to simply represent a `nf Lazy.t` by a pair `env * t` (a delayed computation / a term inside an explicit substitution), and rely on the memoization layer to avoid duplicated computations. This is a cool trick! But it feels weird, because it means that the normalization strategy now depends on memoization. (Previously I thought that I had a standard reduction strategy, with a memoization layer on top for repeated subexpressions. But, thinking more about it, I was probably already wrong.) I will try to implement the strong call-by-need evaluator in this way.

**Comment by Ekdohibs** [comment]
I thought about it as well and arrived to the same conclusion as you. However, I hadn't thought about simply using the memoization itself, and that would work! It even makes a lot of sense if you think of call-by-need as memoizing call-by-name instead of as lazy call-by-value.

**Comment by gasche** [comment]
I have updated the PR to use the trick that replaces the lazy thunks; the code is smarter and shorter. We are back to hashtables, I don't need to complain about `Ident.tbl` anymore.

This is all good fun, but we are still missing a reviewer.

**Comment by Octachron** [comment]
I am available to review the code, but as the person that missed the issue with the initial reduction strategy, I don't think that I am the best reviewer available. Maybe a paired review with @Ekdohibs would be better?

File: typing/ident.mli
Comment (Ekdohibs) [inline typing/ident.mli]: Is this still needed?

Comment (gasche) [reply typing/ident.mli]: Not if we are happy with the latest iteration (without a custom comparison function), but I wasn't sure so I kept the work in history for now. I could still drop this commit. (It would probably be wise to replace `Ident.tbl` by `Map.Make(Ident)` anyway.)

**Comment by gasche** [comment]
Last Tuesday, I had a meeting with @trefis, @voodoos, @Ekdohibs and @Octachron to review this PR. I am now done taking all the review comments into account.

**Comment by Octachron** [comment]
Comparing the compilation time and cmt file size between this PR, #10796, and a version with 4.14 with shape reverted on [a middle-sized set of opam packages](https://github.com/Octachron/ocaml-perfomance-monitoring/blob/master/shape_reduction/full_trace.ml), this PR seems to be the better behaved version: the increase of the filesize and compilation time for the worse/hardest compilation units are much more moderate. Contrarily, for more typical compilation units, this PR does only slightly worse than #10796 .

More precisely, the quantiles for the relative size increase of cmt compared to the reverted version

| quantiles | #10796 | strong call by need |
|---|------------------------| -----------------------------|
| 1% | 1.00087| 1.00122 |
| 10% | 1.00273| 1.00325 |
| 25% | 1.00521| 1.00577 |
| 50% | 1.0098| 1.01219 |
| 75% | 1.01865| 1.02394 |
| 90% | 1.03115| 1.03963 |
| 99% | 1.08271| 1.09222 |
| 99.9% | 6.50957| 1.29952 |
| 99.99% | 37.8402| 2.68921 |

show that we avoid an increase by an order of magnitude of two with the strong call by need implementation in the last millile of compilation units. At the same time, for more than 99% of compilation units, the cmt file size increase is less than 10% which seems fine.

Similarly, the quantiles for the relative increase in total compilation time

| % |  #10796 | strong call by need |
|---|------------------------| -----------------------------|
| 1% | 0.894737| 0.909091 |
| 10% | 1| 1 |
| 25% | 1| 1 |
| 50% | 1| 1 |
| 75% | 1.03846| 1.04348 |
| 90% | 1.09091| 1.1 |
| 99% | 1.21429| 1.25 |
| 99.9% | 1.36364| 1.5 |
| 99.99% | 3.26433| 1.66327 |

shows a more moderate increase for the worse outliers with this PR implementation. And reassuringly, for more than 90% of compilation units the increase of compilation time is less than 10%. This is a more visible increase than the increase in file size that seems still ok.

Thus, once @trefis and me have finished this last round of review, I propose to merge this PR. I will then release an alpha release of 4.14 in order to gather feedback for the larger community while I extend my tests to a larger sets of packages.

**Comment by gasche** [comment]
Thanks for the measurements!

Nitpick: I think that using quantiles to study the outliers is a bit misleading: if there is one outlier that blows up every 10^5 file, then looking at an average over the worst 10^4 files is going to hide one order of magnitude from the blowup. But to the one programmer whose code blows up, the fact that they are a Dirac is no consolation. I think it would be worth adding the *one* encountered worst-case to your analysis.

**Comment by Octachron** [comment]
Checking the data, in this case the last 99.99% quantile contains exactly one point for both tables, which are respectively `irmin.cmt` and `irmin.ml`. But I agree that for the wider test, the very tail of the distribution is also important.

File: testsuite/tests/shapes/recmodules.ml
Comment (trefis) [inline testsuite/tests/shapes/recmodules.ml]: Looks like the functor argument (`A`, defined in the same recursive group) which was previously a `Var` is now a `Leaf`.

File: typing/shape.ml
Comment (trefis) [inline typing/shape.ml]: This already exists as `Ident.Map`.

File: typing/includemod.ml
Comment (trefis) [inline typing/includemod.ml]: As @Octachron already pointed out, this would be nicer as a utility function in the shape module:
```ocaml
val decompose_abs: t -> var * t
```

(I'm the one who originally wrote this, but since you've now taken ownership \o/)

File: testsuite/tests/shapes/recmodules.ml
Comment (trefis) [inline testsuite/tests/shapes/recmodules.ml]: Here too we now have a `Leaf` shape. Note that this is different from what happens in #10796 , where they are kept as variables (cf [here](https://github.com/ocaml/ocaml/pull/10796/files#diff-604a5dcfd9b59dab3398cbbfefa12dce7a5afaf8b6278482351ce6d9288685edL46)).

**Review by trefis — COMMENTED** [review]
I still have to read the reduce code in details, but I left a few comments already.
Something seems off with recursive modules, but I don't know why yet.

**Comment by gasche** [comment]
> Here too we now have a `Leaf` shape. Note that this is different from what happens in #10796 , where they are kept as variables (cf [here](https://github.com/ocaml/ocaml/pull/10796/files#diff-604a5dcfd9b59dab3398cbbfefa12dce7a5afaf8b6278482351ce6d9288685edL46)).

The behavior here comes from this piece of code (in the current version of this PR):

```ocaml
      | Var id ->
          [...] (* first try in the local environment, then: *)
          match Params.find_shape global_env id with
          | res when res = t -> return (NVar id)
          | res ->
              decr fuel;
              reduce env res
          |  exception Not_found ->
             return NLeaf (* avoid loops. *)
          end
```

Before we reviewed together, I had conflated the cases `res when res = t` and `Not_found` to `return (NVar id)` in both cases, but people asked me to revert the `Not_found` behavior as in the pre-PR state, which in fact sends *both* those cases to Leaf instead:

```ocaml
      | Var id ->
          begin try
            let res = Params.find_shape env id in
            if res = t then
              raise Not_found
            else
              decr fuel;
              reduce env res
          with Not_found -> { t with desc = Leaf } (* avoid loops. *)
          end
```

I'm okay with whatever, just tell me what you prefer. `Leaf` erases some information, but note that the behavior of keeping `(N)Var id` results in shapes with unbound identifiers (that are not `Comp_unit` nodes), due to the odd translation of recursive modules.

**Comment by gasche** [comment]
Note: if I understand correctly, the reason why the `Leaf`-introduction behavior does not show up in the test results of your own PR is that the outputs we see in that PR are not actually normalized (some normalization happens on the fly during shape construction, but not the normalization of the fields of recursive-module definitions).

**Comment by Octachron** [comment]
This PR behavior corresponds to moving the reduction of the shape of recursive module from Merlin to the typechecker local reduction. Do we expect that Merlin will try to call the reduction with more fuel? If not, the current behavior seems simpler.

**Comment by gasche** [comment]
Personally I think that:
- We have decided to not look too closely at this recursive-module behavior for now.
- Returning `NVar id` in all these weird cases is the safer choice, because we stop and loose no information that was present in the input term. (So Merlin can do whatever.)
- Longer-term we should either change the translation of recursive modules into shape (in typemod.ml) by adding a proper, explicit rec/fixpoint construct, or we should treat all shape structures as if their fields were mutually-recursive. Either approaches would let us do a better job on recursive modules, but I think they are outside the scope of the present PR and not a 4.14 requirement.

**Comment by Octachron** [comment]
I had forgotten that the local reduction does not use the global environment for recursive modules. Thus I agree in fact that we want to return `NVar id` in all cases to let the hand to  `Merlin`.

**Comment by trefis** [comment]
Let's go with returning `NVar` and improving at a later stage.

**Comment by gasche** [comment]
I pushed a new commit to go back to `NVar` in both cases. This fixed both testsuite differences remarked by @trefis.

File: testsuite/tests/shapes/recmodules.ml
Comment (gasche) [reply testsuite/tests/shapes/recmodules.ml]: (fixed)

File: testsuite/tests/shapes/recmodules.ml
Comment (gasche) [reply testsuite/tests/shapes/recmodules.ml]: (fixed, now as in #10796; I wondered why we don't get `("t", type) -> ...` anymore, this is due to the optimization on full signatures that performs an eta-reduction / avoid expanding into a new structure.)

File: typing/shape.ml
Comment (gasche) [reply typing/shape.ml]: I think you are working on a codebase from the future, my ident.mli has no Map module (unlike Path).

Comment (trefis) [reply typing/shape.ml]: https://github.com/ocaml/ocaml/blob/trunk/typing/ident.mli#L20 , you can grep your codebase from the past, it's already used.

Comment (gasche) [reply typing/shape.ml]: To my defense, it's hard to tell what names are defined in a ML module...

File: typing/includemod.ml
Comment (gasche) [reply typing/includemod.ml]: done in a new commit

(I didn't make the type `Shape.t` (or `Shape.desc`?) abstract, I'm not sure that the code improves when those AST types are more abstract. I'll let other decide what is best and send a PR if they care.)

**Comment by gasche** [comment]
I believe that I have addressed all the recent review comments.

File: typing/includemod.ml
Comment (Octachron) [reply typing/includemod.ml]: I agree that it is not really useful at this point to make `Shape.t` abstract.

**Review by Octachron — APPROVED** [review]
I have reread the code one last time, this looks good to merge now and it definitively better than what you have in the current 4.14 branch (a bout of paranoia made me check that this last version of the PR still lead to measurements comparable to the one I had posted before) .

**Comment by Octachron** [comment]
Extending my measurement on nearly [half of opam packages (that compiles with 4.14)](https://github.com/Octachron/ocaml-perfomance-monitoring/blob/master/shape_reduction/total_trace.ml) confirms that for 99% of compilation units the increase in cmt files is less than 10% (I am still collecting data on compilation time), 


| quantiles | #10796 | Strong call-by-need |
|---|------------------------| -----------------------------|
| 1% | 1.0006| 1.00087 |
| 10% | 1.00272| 1.00357 |
| 25% | 1.00519| 1.00663 |
| 50% | 1.00997| 1.01368 |
| 75% | 1.01921| 1.02865 |
| 90% | 1.03108| 1.04377 |
| 99% | 1.06194| 1.09982 |
| 99.9% | 1.27163| 1.19425 |
| 99.99% | 6.7198| 1.90304 |

Similarly, the increase of cmt files in the ten worst case (out of 12 000 cmt files) is much more reasonable with this PR (and irmin remains an outstanding outlier)

file  | #10796 | Strong call-by-need | quantiles
------------|------------|------------|------------|
 frama-c.24.0:Wp.cmt | 1.45974 | 1.19695 | 99.9195 
irmin-git.2.10.0:irmin_git.cmt |  1.46695 | 1.23853 | 99.9276 |
preface.0.1.0:preface_make__Arrow_apply.cmt |  1.57351 | 1.24677 | 99.9356 |
| frama-c.24.0:Qed.cmt  | 1.90178 | 1.26065  | 99.9437 |
| preface.0.1.0:preface_make__Arrow_zero.cmt  | 1.94462 | 1.27197 | 99.9517
 | cmdliner.1.0.4:cmdliner_term.cmt  | 2.68721 | 1.29952| 99.9598
| preface.0.1.0:preface_make__Arrow_choice.cmt| 3.71465 | 1.36212  | 99.9678 
| preface.0.1.0:preface_make__Arrow_alt.cmt | 6.50957 | 1.46014 | 99.9759 | 
 | preface.0.1.0:preface_make__Arrow_plus.cmt | 6.7198 | 1.90304  | 99.9839
| irmin.2.10.0:irmin.cmt   |  37.8402 | 2.68921 | 99.992

**Pull request merged**
